Don't call `git ls-files` on every invocation of make

dra27 opened the pull request:
This is a more permanently solution than that proposed in https://github.com/ocaml/ocaml/issues/8742#issuecomment-502471968.

An alternative way would be to fall-back to using `find` instead of `git ls-files` if the `.git` directory/file is not present (i.e. the build tree is from a tarball), but that introduces an unnecessary maintenance overhead. With this change, any bugs in the process should only affect developers who are building the `*.ast` files.

A fix will be wanted on 4.08 and 4.09.

dra27 commented:
@Armael pointed out in https://github.com/ocaml/ocaml/issues/8742#issuecomment-502854380 that while I'd fixed `partialclean` (which would affect `make bootstrap`) that the `build-all-asts` target was the underlying problem. I've pushed an extra commit to deal with this.

gasche commented:
Apologies for introducing the issue in the first place.

From a high-level point of view: these are for-development features that are causing trouble. Instead of rewriting them in clever ways to not depend on development tools, could we instead move them to a `Makefile.dev` file that is not in the build/install path, and train our contributors to use `make -f Makefile.dev` on the occasion?

xavierleroy commented:
+1 for moving developers-only features outside of the main Makefile.

dra27 commented:
I'd be completely behind that approach too, if you're happy with it! There's a possible neater option which would not require the training - the inclusion of development rules could be enabled via `configure` either manually or automatically (i.e. either add `--enable-dev-mode` or assume that if we're in a git clone then development rules should be enabled - there's already stuff which does that for CI). That's then a matter of a guarded `include Makefile.dev` line in the toplevel `Makefile`

gasche commented:
Yes, I definitely support moving these rules out of the way, even if it costs a small amount of convenience on usage.

(Not convinced about special configure-time arguments; it makes the build system more complex, and I never remember to use the special options anyway, I just `./configure` most of the time. These extra rules have to be mentioned in HACKING or somewhere anyway, so I don't think using a separate makefile is a problem.)

We could also consider using the same approach for `Makefile.menhir`, instead of including it in the main Makefile.

@dra27: I'm would not want to overload your boat with many things to do for consistency. Please do the minimal thing that you think is a good solution, and we can grow these separate Makefiles later. In fact, I would understand if you expected me to clean my own mess, even though I'm very happy to let you handle it.

dbuenzli commented:
> (Not convinced about special configure-time arguments; it makes the build system more complex, and I never remember to use the special options anyway, I just `./configure` most of the time. 

Well the same can be said of make ("I usually simply `make` and will never remember I have to `make -f Makefile.dev`). Why not simply include `Makefile.dev` only if there's a `.git` directory at the root ? Presumably distribution archives do not have that directory.

dra27 commented:
@dbuenzli - yeah, that's simpler than messing around in `configure` (it's how flexdll bootstrap is activated too). I think including the rules on any development box is also a good way of preventing bitrot - the tarball version will always be being tested by opam switches.

I'll push something in a bit - overloaded VMs here at the moment :)

dra27 commented:
There we go - I think this is simplest. If a `.git` directory (or file, allowing for worktrees) exists, then `Makefile.dev` is `include`d in the toplevel `Makefile`. So developers get all the rules, but release tarballs won't.

dra27 committed:
Move AST testing rules to Makefile.dev

Makefile.dev is only included if a Git clone is detected.

dra27 committed:
Tweak build-all-asts target

make evaluates macros in target specifications when the Makefile is
read, which meant that the build-all-asts target caused `git ls-files`
to be called on every invocation of make. This tweaks it to be a
recursive call to make which puts the $(AST_FILES) macro in the recipe
where it is only evaluated when the target is requested.

dra27 commented:
I've cherry-picked the original tweak to `build-all-asts` to `Makefile.dev` - `git ls-files` is fast, but there's no point running it every time a developer runs `make`, either! The original change to `partialclean` I've dumped, as that's adds far to weight simply to optimise `make coreboot`.

dra27 committed:
Update HACKING.adoc

gasche commented on Makefile.dev:
```diff
@@ -0,0 +1,48 @@
+#*                                                                        *
+#**************************************************************************
+
+# Developer-only rules
```
maybe "included in Makefile when a .git repository is detected"?

gasche reviewed (approved):
I like the new approach, and I agree with the conditional-include logic being a sweet spot (thanks @dbuenzli for the suggestion).

dra27 committed:
Tweak

dra27 commented:
I'll squash and cherry-pick once CI completes the hopefully obvious confirmation
